### PR TITLE
Update translator protocol

### PR DESCRIPTION
**Pull request by soderluk**
Add environment variables for private/public keys
Add support for validating headers
Add support for signing data payload
Update Pipfile
Update README
Update CHANGELOG
Update API RAML specification
Remove shared secret setting
Replace generate_signature and validate_signature with generate_signed_data and validate_signed_data functions

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -21,18 +21,60 @@ You need the following installed on your computer:
 
 ## Implementing a translator
 
-In `settings.py` there's the `SHARED_SECRET` that you have to set. The shared 
-secret is shown once, when the data product is created in the Platform Of Trust
-Product API.
+Add the Platform Of Trust public key to the `settings.py` `POT_PUBLIC_KEY`
+environment variable. This public key is used for validating the signature
```
Comment (lietu) [inline README.md]: Python variables are not the same as environment variables :)

File: README.md [hunk 2]
```diff
@@ -21,18 +21,60 @@ You need the following installed on your computer:
...
+line.
+
+# Generating private and public keys
+
+To generate the PEM keys used for the signing of data:
+
+    ssh-keygen -t rsa -b 4096 -q -N "" -m PEM -f RS256.key
+    openssl rsa -in RS256.key -pubout -outform PEM -out RS256.key.pub
+
+Base64 encode the public and private keys
```
Comment (lietu) [inline README.md]: There's no reason provided here for doing so, and this practice is really only specific to Kubernetes secrets, so I would remove it.

File: app/services.py [hunk 3]
```diff
@@ -12,15 +16,60 @@ class Translator(object):
...
         :return: The data to be sent back to the Data Broker API.
         :rtype: dict
         """
-        return {}
+        created_at = app.utils.rfc3339()
+        # Get the data for the translator response here.
+        data = {}
+        signature = self.sign_data(data, created_at)
+        # Set the @context, @type, signature.type, signature.creator
+        # The signature.type MUST be same as defined in the data product
```
Comment (lietu) [inline app/services.py:37]: Maybe it would be helpful to mention this combination should match *one* of those defined for the data product.

File: app/services.py [hunk 4]
```diff
@@ -12,15 +16,60 @@ class Translator(object):
...
+        :return: The signature.
+        :rtype: str
+        """
+        # Include the created at time in the data.
+        data['__signed__'] = created_at
+        # Todo: Use the corresponding private key that was defined in the
+        #  data product organizationPublicKeys[].url
+        signature = app.utils.generate_signed_data(data, settings.PRIVATE_KEY)
+        # We don't want the "metadata" to be included in the response.
+        del data['__signed__']
```
Comment (lietu) [inline app/services.py]: I was thinking, maybe 

```python
from copy import copy
sign_data = copy(data)
sign_data["__signed__"] = created_at
signature = app.utils.generate_signed_data(sign_data, settings.PRIVATE_KEY)
```

would be *slightly* better - no chance the original data is modified in any way

File: app/tests/test_controllers.py [hunk 5]
```diff
@@ -47,35 +54,48 @@ class TestTranslatorController(BaseTestCase):
     """Tests the translator controller."""
 
     _body = {
-        "timestamp": "2018-11-01T12:01:01Z",
-        "productCode": "product-1",
-        "parameters": {
-            "name": "The name",
+        'timestamp': '2018-11-01T12:01:01Z',
```
Comment (lietu) [inline app/tests/test_controllers.py:57]: Nothing wrong with double-quotes in Python, why this change?

File: app/utils.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+X_POT_SIGNATURE = 'x-pot-signature'
+X_POT_TOKEN = 'x-pot-token'
+X_POT_APP = 'x-pot-app'
+
+SUPPORTED_HEADERS = {
+    X_POT_SIGNATURE: {
+        'required': True,
+    },
+    X_POT_TOKEN: {
+        'required': False,
```
Comment (lietu) [inline app/utils.py:27]: If this item is never used for anything if it's not required, why have this data structure like this instead of e.g.

```python
REQUIRED_HEADERS = (X_POT_SIGNATURE, X_POT_APP,)
```

File: app/utils.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    valid_signature = validate_signed_data(
+        body,
+        base64.b64decode(headers[X_POT_SIGNATURE].encode('utf-8')),
+        settings.POT_PUBLIC_KEY
+    )
+    if not valid_signature:
+        raise ValidationError('Signature validation failed.')
+
+    # Todo: validate timestamp as well.
```
Comment (lietu) [inline app/utils.py:93]: We're so close to release that we should probably not leave todos like this in :)

Comment (lietu) [reply app/utils.py:93]: Could e.g. use something like fail if timestamp is older than 60 seconds or some such - to allow for some clock drift but not much.

File: app/utils.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    if not isinstance(body, (str, dict)):
+        raise RuntimeError(
+            f'Invalid body type. Must be `str` or `dict`, got {type(body)}')
+
+    # Create the hash from the body dict.
+    if isinstance(body, dict):
+        body_hash = json.dumps(
+            body,
+            sort_keys=True,
+            indent=None,
```
Comment (lietu) [inline app/utils.py:116]: I thought we agreed on an indentation of 2 spaces or some such for all of these?

File: app/utils.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+            separators=(',', ': ')
```
Comment (lietu) [inline app/utils.py:117]: Should the `','` also have a space?

File: app/utils.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    """Generates signature for payload.
+
+    :param payload: The payload to sign.
+    :type payload: str|dict
+    :param private_pem: The private key used for signing.
+    :type private_pem: str
+    :return: The signature.
+    :rtype: str
+    """
+    payload_hash = get_signature_payload(payload)
```
Comment (lietu) [inline app/utils.py]: The contents of this variable is not a hash

File: app/utils.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        body_hash = json.dumps(
+            body,
+            sort_keys=True,
+            indent=None,
+            separators=(',', ': ')
+        ).strip()
+    else:
+        body_hash = body.strip()  # Strip white space from start and end.
+
+    return body_hash
```
Comment (lietu) [inline app/utils.py]: The contents of this variable is not a hash

File: app/utils.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    :param payload: The payload to validate.
+    :type payload: str|dict
+    :param signature: The signature to validate.
+    :type signature: str
+    :param public_pem: The public key used for validating.
+    :type public_pem: str
+    :return: True if signature valid, False otherwise.
+    :rtype: bool
+    """
+    payload_hash = get_signature_payload(payload)
```
Comment (lietu) [inline app/utils.py]: The contents of this variable is not a hash

File: docs/api.raml [hunk 13]
```diff
@@ -39,17 +39,17 @@ types:
       X-PoT-Signature:
         description: |
           A HMAC-SHA256 generated signature of the payload.
-          E.g. HMAC-SHA256(sharedSecret, body)
+          E.g. HMAC-SHA256(<pot private key>, <body>)
```
Comment (lietu) [inline docs/api.raml]: It is also HMAC-RSA256 now, right?

File: docs/api.raml [hunk 14]
```diff
@@ -81,13 +81,20 @@ types:
                   }
                 }
             signature:
-              description: Signature information
+              description: |
+                Signature information. The signature is generated based on the
+                `data` object with the `created` value added to the key
+                `__signed__`. The `__signed__` is NOT returned in the data, but
+                MUST be added when validating the signature.
+                The `type` MUST match the data product `organizationPublicKeys.type`
```
Comment (lietu) [inline docs/api.raml]: Again `one of` should probably go here

File: settings.py [hunk 15]
```diff
@@ -28,8 +28,91 @@
...
+WEBndfq5T0+pnBDJXISxzbe7ZboNH5VEKi4zzDcqJOcqexhZSimd21YHVRhcUhya
++G3JCzb3kxK7dROPSA/i0S+TcGyrHz177HbYT7aKs6zoaJsInfMSirEr2wqhVo4G
+Ra4hCUoqK+Trb26K6jKd98uMQrZL6EigobDGs7qfILCSWGStB7qzax35NqrJy9wz
+4yb3Sknd9ujFOTvPNgaMwmv0PC77dAaP46FG+jPS3M/PKAuP0rw2wjjdJRv8eBEB
+FKDg6EtsfQ70q1q7Nm/B4R/SvFOeQWSzHjBE2ssPmDapfK1CjpXTSys4Vvf4XIov
+fZbnsMS87zXeN0TnvkGOspMCAwEAAQ==
+-----END PUBLIC KEY-----'''
+
+# Private key used for signing. THIS IS ONLY FOR TESTING PURPOSES.
+PRIVATE_KEY = '''-----BEGIN RSA PRIVATE KEY-----
```
Comment (lietu) [inline settings.py]: I was actually wondering - could we for Travis tests use env variables set in Travis and like a `settings.travis.py`? Would leave less of a chance for people to actually copy these to their own things.

Applies to all 3 variables really.

**Review by lietu — CHANGES_REQUESTED** [review]
Found some things to think about

File: app/tests/test_controllers.py (hunk 5 above)
Comment (soderluk) [reply app/tests/test_controllers.py:57]: I try to keep things consistent, and I've been using single quotes practically everywhere. Just a style thing.

File: README.md (hunk 1 above)
Comment (soderluk) [reply README.md]: Removed `environment`.

File: app/services.py (hunk 3 above)
Comment (soderluk) [reply app/services.py:37]: Will make a note of that.

File: app/services.py (hunk 4 above)
Comment (soderluk) [reply app/services.py]: Good thinking.

File: README.md (hunk 2 above)
Comment (soderluk) [reply README.md]: OK.

File: app/utils.py (hunk 11 above)
Comment (soderluk) [reply app/utils.py]: True.

File: app/utils.py (hunk 12 above)
Comment (soderluk) [reply app/utils.py]: True.

File: docs/api.raml (hunk 13 above)
Comment (soderluk) [reply docs/api.raml]: It is, yes.

File: docs/api.raml (hunk 14 above)
Comment (soderluk) [reply docs/api.raml]: Yup.

File: settings.py (hunk 15 above)
Comment (soderluk) [reply settings.py]: Probably. Could be a good solution.

File: README.md [hunk 16]
```diff
@@ -63,18 +61,6 @@ There is an `invoke` task for running tests, run `pipenv run invoke test`.
...
 line.
 
-# Generating private and public keys
-
-To generate the PEM keys used for the signing of data:
-
-    ssh-keygen -t rsa -b 4096 -q -N "" -m PEM -f RS256.key
-    openssl rsa -in RS256.key -pubout -outform PEM -out RS256.key.pub
-
-Base64 encode the public and private keys
```
Comment (lietu) [inline README.md]: I was referring just to the Base64 bit when I suggested removing it :)

Comment (soderluk) [reply README.md]: Ah! That wen past me :)

File: settings.py (hunk 15 above)
Comment (soderluk) [reply settings.py]: Moved the keys to travis environment variables, and added settings.travis.py

**Review by lietu — APPROVED** [review]
Looks good!
